src/shogun/machine/GLM.h

```diff
@@ -0,0 +1,271 @@
+ *  alpha = 0 and defaults to the lasso penalty when alpha = 1.
+ *
+ * */
+class GLM : public RandomMixin<IterativeMachine<LinearMachine>>
+{
+ 	public:
+		friend class GLMCostFunction;
+		
+		/** default constructor */
+		GLM();
+
+ 		GLM(GLM_DISTRIBUTION distribution, float64_t alpha, float64_t lambda,
+		 	float64_t learning_rate, int32_t max_iterations,
+			 float64_t tolerance, float64_t eta);
+
+ 		/** destructor */
+ 		~GLM() override {}
+
+		/** apply linear machine to data
+		 * for regression problem
+		 *
+		 * @param data (test)data to be classified
+		 * @return classified labels
+		 */
+		std::shared_ptr<RegressionLabels> apply_regression(std::shared_ptr<Features> data=NULL) override;
+
+ 		/** Returns the name of the SGSerializable instance.  It MUST BE
+ 		 *  the CLASS NAME without the prefixed `C'.
+ 		 *
+ 		 * @return name of the SGSerializable
+ 		 */
+ 		const char* get_name() const override { return "GLM"; }
+
+		void set_tau(SGMatrix<float64_t> tau);
+
+		SGMatrix<float64_t> get_tau();
+
+	protected:
+
+		void init_model(const std::shared_ptr<Features>& data);
+		
+		void iteration() override;
+
+	protected:
+
+		/** Distribution type */
+		GLM_DISTRIBUTION distribution = POISSON;
+
+		/** a threshold parameter that linearizes the exp() function above eta. */
+		float64_t m_eta = 2.0;
+
+		/** regularization parameter :math:`\\lambda` of penalty term. */
+		float64_t m_lambda = 0.1;
+
+		/** the weighting between L1 penalty and L2 penalty term of the loss function. */
+		float64_t m_alpha = 0.5;
+
+		/** convergence threshold or stopping criteria. Optimization loop will stop when relative change
+		 * in parameter norm is below the threshold. */
+		float64_t m_tolerance = 1e-6;
+
+		bool m_compute_bias = true;
+
+		std::shared_ptr<GradientDescendUpdater> m_gradient_updater;
+
+		std::shared_ptr<ConstLearningRate> m_learning_rate;
+
+		std::shared_ptr<ElasticNetPenalty> m_penalty;
+
+		std::shared_ptr<GLMCostFunction> m_cost_function;
+};
+
+class GLMCostFunction
+{
+public:
+
+	friend class GLM;
+
+	GLMCostFunction()
+	{
+
+	}
+
+	virtual float64_t get_cost()
+	{
+		//TODO
+		return 0.0;
+	}
+
+	virtual SGVector<float64_t> get_gradient(const SGMatrix<float64_t> X, const SGVector<float64_t> y,
+												const SGVector<float64_t> w, const float64_t bias,
+												float64_t lambda, float64_t alpha, bool compute_bias,
+												float64_t eta, GLM_DISTRIBUTION distribution)
+	{
+		auto n_samples = y.vlen;
+	
+		auto z = compute_z(X, w, bias);
+		auto mu = non_linearity(z, compute_bias, eta, distribution);
+		auto grad_mu = gradient_non_linearity(z, eta, distribution);
+
+		SGVector<float64_t> grad_w(w.vlen);
+		SGVector<float64_t> a;
+		switch (distribution)
+		{
+		case POISSON:
+			{
+			a = linalg::element_prod(y, grad_mu);
+			for(auto i : range(y.vlen))
+				a[i] /= mu[i];
+
+			auto prod1 = linalg::matrix_prod(SGMatrix(grad_mu), X, true, true);
+			auto prod2 = linalg::matrix_prod(SGMatrix(a), X, true, true);
```

Comment:
@karlnapf woudn't it be better to also write the one line formula for this as a comment?

Reply:
+1

Reply:
Okay, I'll add this.